### DISCUSSION TITLE
How to change source of an image tag in XAML from C# dynamically?

### DISCUSSION THREAD

**Discussion by righthere** [AvaloniaUI/Avalonia#8394] [General]
I tried add an image, not from resources in Assets, but from anywhere to display an image in a C# Avalonia application in a XAML Image tag, to create an image viewer or photo sorting application. But I really don't know at this point how to do that. I tried and searched around for a lot of time, to have an image dynamically loaded not as an embedded resource but as an external one. I have successfully loaded from assets with avares:

 `<Image Source="C:/Users/pc/Documents/imagetoview.jpg" Width="128" Stretch="Uniform" />`
Those full url are loaded directly from ouside the embedded image resources, so I know now that Avalonia is capable to load images at runtime, but here comes the C# part, that I don't know how to solve:

As far as I understood this is the procedure, first bind with C# a variable in XAML

`<Image Source="{Binding ImageToView}"`
Then in C# add the variable ImageToView

`public Bitmap ImageToView;`
After that in ModelView constructor (or in a function) change the image with another bitmap, I tried:

```
var bitmap = new Bitmap("C:/Users/pc/Documents/anotherimagetoview.jpg");
var imgStream = new MemoryStream();
bitmap.Save(imgStream);
imgStream.Seek(0, SeekOrigin.Begin);
ImageToView.Source = new Bitmap(imgStream);
```
And another method:

```
var assetsWhat = AvaloniaLocator.Current.GetService<IAssetLoader>();
ImageToLoad.Source = new Bitmap(@"C:/Users/pc/Documents/anotherimagetoview.jpg");
This one with converter got from this place: https://github.com/VitalElement/AvalonStudio.Shell/blob/master/src/AvalonStudio.Shell.Extensibility/Converters/BitmapValueConverter.cs
```

Or (this operator gives me errors):

`Image imageLoad = this.FindControl<Image>("TheImage");`
And:

`Image? imageLoad = ((IControl)hostWindow.GetVisualChildren().FirstOrDefault())?.FindControl<Canvas>("MyImage");`
No methods worked, are all the various methods related to this topic available, that is surprising because I solved with this frameworks more advanced issues, but I for sure have missed something because it's not possible that is so difficult to load a simple image in Avalonia.

This took me days to search everywhere for a solution, but I haven't found one, so I hope that someone had my same issues and has already solved. Thanks in advance!

**Comment by maxkatz6** [comment]
This should be enough:
```csharp
ImageToView = new Bitmap("C:/Users/pc/Documents/anotherimagetoview.jpg");
```

Or if you want to use stream:

```csharp
using var fileStream = File.OpenRead("C:/Users/pc/Documents/anotherimagetoview.jpg");
ImageToView = new Bitmap(fileStream);
```

> public Bitmap ImageToView;

If you want to bind to a property, it MUST be a **property**. ImageToView is a field in your case. I suppose it's the root of your problem here.

> var assetsWhat = AvaloniaLocator.Current.GetService<IAssetLoader>();

IAssetLoader is needed when you want to load bitmap from Avalonia resources (when it embedded inside of the dll/exe).

> Or (this operator gives me errors): Image imageLoad = this.FindControl<Image>("TheImage");

You didn't show what error it gives. If image is visible in the namescope of the user control/window, that code will work.

You also can use https://github.com/AvaloniaUI/Avalonia.NameGenerator to access controls from code behind, similarly as in WPF/UWP. With newer templates it is included by default.

**Comment by righthere** [comment]
Hi! Thanks for your reply!
I tried the code but wasn't working, then I remembered that there are other things to add like getters and setters, indeed I added this to declaration:
`public Bitmap ImageToView { get; set; }`
And that's worked, finally I changed the image on load :-)
Now, I wanted to modify all programmatically to be sure, and of course didn't work.
I partially added the official guide for bitmap changes listener (I guess) on Avalonia website to build a music store app [here](https://docs.avaloniaui.net/tutorials/music-store-app/displaying-images):
```
private Bitmap? ImageToLoad;
public Bitmap? ImageToLoadPublic
 {
     get => ImageToLoad;
     private set => this.RaiseAndSetIfChanged(ref ImageToLoad, value);
  }
```
Then added a XAML button to test if on click the image would change (and set the Image binding to ImageToLoadPublic to listen on changes):
```
<StackPanel>
      <Button Width="100" Margin="250,0,0,0"
          Foreground="White" Background="BLue"
          Command="{Binding ChangeImageCommand}"
        >Change Image
        </Button>
 </StackPanel>
```

And in C# again the binding logic:
```
public ReactiveCommand<Unit, Unit> ChangeImageCommand { get; set; }

 ChangeImageCommand = ReactiveCommand.Create(() =>
 {
      TestSound.Play(); //this is to test that the button has been bound 
       using var fileStream = File.OpenRead("C:/Users/pc/Documents/anotherimagetoview.jpg");
       ImageToLoad = new Bitmap(fileStream);
 });
```
The image is still loaded on start, but when I click the XAML button the image is not changed. I know that the button command is working because I play a little sound effect to test.

Thanks in advance!

**Reply by timunie** [reply]
ImageToLoad is the private field and the UI will not be notified. Either call RaisePropertyChanged manually on ImageToLoadPublic or assign the new Bitmap to the public property.

**Reply by righthere** [reply]
Hi! I changed ImageToLoadPublic and ImageToLoad all to public and nothing changed.
I don't know why the change is notified even in the XAML designer (seems that before compilation the image is parsed), but when I try to do the same thing but with a command the change fails.
Can you please explain how to do with my code sample above? Thanks!

**Reply by jp2masa** [reply]
The command should set the property, not the field, so that `RaiseAndSetIfChanged` in the setter is called, and the UI can update accordingly. So instead of `ImageToLoad = new Bitmap(fileStream);` you should have `ImageToLoadPublic = new Bitmap(fileStream);`.

**Reply by righthere** [reply]
Hey it works :-)
Can you explain why ImageToLoadPublic worked? I don't understand this passage. So the listener, when a set event will occur will update the ImageToLoad bitmap too? Are they bound with the "get" setter or ImageToLoadPublic?

**Reply by jp2masa** [reply]
When you do `ImageToLoad = new Bitmap(fileStream);`, you're updating the field and nothing more, there's no way for the UI to know it has to be updated. When you do `ImageToLoadPublic = new Bitmap(fileStream);`, it will call the setter of `ImageToLoadPublic`, which calls `this.RaiseAndSetIfChanged(ref ImageToLoad, value);`, which sets `ImageToLoad` and invokes `PropertyChanged`, notifying the UI that it has to update.

**Reply by righthere** [reply]
Perfect! Thanks again!
